components/data-bridge/org.wso2.carbon.databridge.agent/src/main/java/org/wso2/carbon/databridge/agent/endpoint/DataEndpoint.java

```diff
@@ -278,22 +279,34 @@ public void run() {
                     publish();
                 } catch (UndefinedEventTypeException ex) {
                     log.error("Unable to process this event.", ex);
+                    semaphoreRelease();
                 } catch (Exception ex) {
                     log.error("Unexpected error occurred while sending the event. ", ex);
+                    semaphoreRelease();
                     handleFailedEvents(this.events);
+                } catch (Throwable t) {
+                    //There can be situations where runtime exceptions/class not found exceptions occur, This block help to catch those exceptions.
+                    //No need to retry send events. Deactivating the state would be enough.
+                    log.error("Unexpected error occurred while sending events. ", t);
+                    semaphoreRelease();
+                    deactivate();
                 }
             } catch (DataEndpointException e) {
                 log.error("Unable to send events to the endpoint. ", e);
+                semaphoreRelease();
```

Comment:
why don't we release in finally

Reply:
As it is released normally when the thread completes execution. Since runtime exception is catched and handled, where the semaphore is release explicitly